### PR TITLE
[FLINK-13417][build] Bundle Zookeeper 3.5 in opt 

### PR DESCRIPTION
**Pull request by zentol**
Based on #11143.

With this we ship 2 versions of Zookeeper (3.4.10, 3.5.6, as defined by flink-shaded) in the distribution.
ZK 3.4 is in /lib, while ZK 3.5 is in /opt.
Users can control which ZK version is used by switching the jars.

The added `zk-3.5` profile is used for CI purposes. It controls which ZK version is used during tests and bundled in lib/opt respectively. It also bumps `curator.version`, which by now is only used for curator-test, to version 4. Version 2 is still required for testing against ZK 3.4 .

### CODE REVIEW

**Comment by StephanEwen** [comment]
Looks good in general.

I am wondering if we could avoid the extra maven profile. With each additional profile, it becomes a tad bit harder to understand/maintain.
Could the tests (at least the end to end tests) not simply copy the jar from opt to lib, rather than relying on a build profile?

**Comment by TisonKun** [comment]
LGTM.

A possible follow-up is separating `curator.version` & `curator.test.version`. zk 3.4 works well with curator 4.x except when testing it requires curator-test 2.x, which is another version than curator-framework and so on.

**Comment by zentol** [comment]
@StephanEwen you still want to use zk 3.5 during IT cases / tests.

**Comment by zentol** [comment]
@TisonKun As noted in the PR description, `curator.version` is only used for curator-test, nothing else. It does not affect the runtime curator version, which is also 4.X.Y.

**Comment by StephanEwen** [comment]
> @StephanEwen you still want to use zk 3.5 during IT cases / tests, and having the logic in the build system is more maintainable than custom code in the e2e scripts.

Is the plan to execute several tests with both ZK versions? Could we then have the profile in the project that has this dependency, rather than in the root pom?

**Comment by zentol** [comment]
The profile modifies `curator.version` which is used in 6 different modules, along with `zookeeper.version` which is used in the dependencyManagement entry for `org.apache.zookeeper` which probably also appears in various placed in the dependency graph.

In other words the root pom is pretty much the only place to put it.

**Comment by TisonKun** [comment]
@zentol Thanks for your explanation. I get it.

**Comment by StephanEwen** [comment]
I see. I guess under the current setup, that approach is fair enough.

It would be nice to keep working on a bit better encapsulation, because this contributes to the increasing "unwieldyness" of the build system.

Concretely, for the ZK / Curator case: Why is that needed in 6 modules? Should it not be only needed in `flink-runtime` for our HA Services? If the same think is used in `flink-mesos`, don't we implicitly require that the Mesos ZK and the Flink HA ZK are the same version?

In a clean setup, I would imagine that we have
  - a `flink-ha-services` module with the HA Services abstraction
  - two different sub-modules for ZK 3.4 (`flink-ha-services-zk-3_4`) and ZK 3.5 (`flink-ha-services-zk-3_5`), with the code and the tests.
  - Maybe most of the code/tests is taken from a shared `flink-ha-services-zk-base` module.
  - Each module does its tests for the ZK HA integration. No dependency change needed in any other module to test cross versions.
  - Possibly one end-2-end tests per version, but those should work "like users" meaning taking the relevant jar from the `/opt` folder.
  -  In that approach, we also have the flexibility to implement the operations differently in different versions, like using "container" nodes in ZK 3.5, transactions, etc.

**Comment by StephanEwen** [comment]
To avoid confusion, I am +1 to merge this, but would suggest to think in the way of the above comments for the future.

**Comment by zentol** [comment]
One example for `curator-test` being used outside of `flink-runtime` is `flink-tests`, which runs some Zookeeper IT cases. For these the curator-test version has to correspond to the version in flink-runtime.
AFAIK some other modules also use curator-test but the flink-runtime ZK may not necessarily be involved here.
Ultimately I don't know what most of these modules do with Zookeeper, but I understood the JIRA to add ZK 3.5 support for _all modules_, not just flink-runtime.

I had an offline chat with Till about this PR.
He raised a good argument in that the "global ZK version" approach is flawed as any module bundling ZK (apart from ZK) is hard-wired at compile-time to that version.
Additionally, it shouldn't be required to actually test against both ZK 3.4 and 3.5 . So long as use 3.4 as the default curator should ensure that we can talk to both versions.
Simply put we propose to the drop the ZK 3.5 profile, only run tests against ZK 3.4, and possible adjust an E2E test to run against 3.5, with the added benefit of actually testing how users would switch between versions.

**Comment by zentol** [comment]
I agree that in the long term we should just have different HA server implementations that are loaded as plugins (at which point we can drop the shading again).

**Comment by zentol** [comment]
ping @StephanEwen

**Comment by StephanEwen** [comment]
> Simply put we propose to the drop the ZK 3.5 profile, only run unit/it tests against ZK 3.4, and possible adjust an E2E test to run against 3.5, with the added benefit of actually testing how users would switch between versions.

I think that proposal sounds good.
